### ISSUE TITLE
kubelet client refactoring: optimization and global variable

### ISSUE THREAD

**Issue by dougbtv** [k8snetworkplumbingwg/multus-cni#673]
@aojea has recommended a few fixes that look to be solid improvements...

1. We're using [these global variables](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/kubeletclient/kubeletclient.go#L23-L26) which is a bad practice, additionally they're being mutated in the kubeletclient_test.go package, and there are some tests which may be running in such a way as to flake.

2. The https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/kubeletclient/kubeletclient.go#L44 `util` package is really only being used for `util.LocalEndpoint()` which is really just a couple methods that manipulate strings, found here @ https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cri/remote/util/util_unix.go#L82-L128 -- we could omit having to include the whole package and instead include these functions.

**Comment by s1061123** [comment]
I agree first comment at all, but a bit wondering about second because it might be changed from 'k8s.io/kubernetes/pkg/kubelet/util' side, we need to track these changes everytime...

**Comment by aojea** [comment]
> I agree first comment at all, but a bit wondering about second because it might be changed from 'k8s.io/kubernetes/pkg/kubelet/util' side, we need to track these changes everytime...

You have the same problem now, if they change and you don't update the module, since modules are pinned to a specific version. 
The only difference is the mechanics ...
The benefits is huge because you can avoid vendoring all the kubernetes repo

**Comment by aojea** [comment]
Do you have tests that covers this possible out of sync?

**Comment by s1061123** [comment]
@aojea but still we use k8s.io/kubernetes/pkg/kubelet/apis/podresources even though we remove k8s.io/kubernetes/pkg/kubelet/util hence we cannot remove completely k8s repo yet. If we completely remove kubelet code then it could be nice.

**Comment by s1061123** [comment]
> Do you have tests that covers this possible out of sync?

I'm not clear that. Could you please let me know the details?

**Comment by s1061123** [comment]
@aojea so you mean that use 'k8s.io/kubelet' instead of 'k8s.io/kubernetes/pkg/kubelet'?

**Comment by s1061123** [comment]
@zshi-redhat any idea for that?

**Comment by aojea** [comment]
> @aojea so you mean that use 'k8s.io/kubelet' instead of 'k8s.io/kubernetes/pkg/kubelet'? If so, it could be better, however current code not only depends on api code also files in 'k8s.io/kubernetes/pkg/kubelet/apis/podresources/', which is not in 'k8s.io/kubelet'.

Yeah, the only code depending on 
k8s.io/kubernetes/pkg/kubelet/apis/podresources/ is the one that creates the client, I've asked today some kubelet folks to check if they can move that to stage

> I'm not clear that. Could you please let me know the details? At least, test does not failed so everything should be in sync

Yeah, so the scenario is this, kubelet changes the socket path upstream, how do you detect that? 
It can happen today, we vendor Kubernetes 1.14 and current version is 1.21, how do we know we need to update?

**Comment by dougbtv** [comment]
re: "_Do you have tests that covers this possible out of sync?_"

> > I'm not clear that. Could you please let me know the details? At least, test does not failed so everything should be in sync
> 
> Yeah, so the scenario is this, kubelet changes the socket path upstream, how do you detect that?
> It can happen today, we vendor Kubernetes 1.14 and current version is 1.21, how do we know we need to update?

Ah ha got it, thanks. Yes, good consideration. We do have an e2e test which uses latest stable k8s, so we should catch this in a reasonable amount of time.

**Comment by s1061123** [comment]
Create another issue to track to remove 'k8s.io/kubernetes', https://github.com/k8snetworkplumbingwg/multus-cni/issues/675

This issue only address global variable.
